src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -804,25 +817,33 @@ async Task RawOpen(SslMode sslMode, NpgsqlTimeout timeout, bool async, Cancellat
                     var checkCertificateRevocation = Settings.CheckCertificateRevocation;
 
                     RemoteCertificateValidationCallback? certificateValidationCallback;
-                    if (sslMode == SslMode.Prefer || sslMode == SslMode.Require)
+
+                    if (Settings.TrustServerCertificate && sslMode is SslMode.Allow or SslMode.VerifyCA or SslMode.VerifyFull)
+                        throw new ArgumentException(NpgsqlStrings.CannotUseTrustServerCertificate);
+
+                    if (UserCertificateValidationCallback is not null)
+                    {
+                        if (sslMode is SslMode.VerifyCA or SslMode.VerifyFull)
+                            throw new ArgumentException(string.Format(NpgsqlStrings.CannotUseSslVerifyWithUserCallback, sslMode));
+
+                        if (Settings.RootCertificate is not null)
+                            throw new ArgumentException(string.Format(NpgsqlStrings.CannotUseSslRootCertificateWithUserCallback));
+
+                        certificateValidationCallback = UserCertificateValidationCallback;
+                    }
+                    else if (sslMode is SslMode.Prefer or SslMode.Require)
                     {
+                        if (isFirstAttempt && sslMode is SslMode.Require && !Settings.TrustServerCertificate)
```

Comment:
What's the reason for not validating this at `NpgsqlConnectionStringBuilder.Validate`, but here? You even have to add `isFirstAttempt` to make it work...

Reply:
The point is to allow Require without TrustServerCertificate, but only if UserCertificateValidationCallback is set. We can't validate this in NpgsqlConnectionStringBuilder since UserCertificateValidationCallback is on NpgsqlConnection, and it seemed natural here after the UserCertificateValidationCallback block above. Note that at some point we'll remove the requirement to explicitly set TrustServerCertificate=true with Require, and then this check (and `isFirstAttempt`) go away. Makes sense?

Also, because all this is a bit tricky and complex, it seems like a good idea to have *all* the SSL-related validation/config in one place...

Reply:
@roji Here if `TrustServerCertificate == false` and `sslMode == SslMode.Require` you imply `TrustServerCertificate == true` on the second attempt because of `certificateValidationCallback = SslTrustServerValidation` below.

I suggest removing `isFirstAttempt &&` from the `if` condition, because then the value of `TrustServerCertificate` does not make sense with `sslMode` set to `SslMode.Require`

Reply:
@slavanap I'm not sure you're correct: the point here is, if it's the first attempt (that is, we connect with the `SslMode` specified by the user), we do throw an exception. But if it's the second attempt (which happens if there was an error while authenticating), we pass `SslMode.Require` if and only if the default `SslMode` is `Allow`, which does imply `TrustServerCertificate=true`.

Reply:
@vonzshik I see condition that allows Prefer or Require mode here https://github.com/npgsql/npgsql/pull/4311/files/b8feab23a62183050779a09b687253c2185c66ac#diff-b45ed9d534512c2b95cd6aafbdceff63d87439bf00b57a605d637cc64785e438R834

I see condition which I've commented on check and throw an exception for Require mode only if firstAttempt is true. Thus, it won't throw on second attempt for Require mode and will disregard TrustServerCertificate / will imply Prefer SSL mode.

Reply:
Yes, because `Require` is passed for the second attempt only if the default mode is `SslMode.Allow`. For `Allow` it doesn't make sense to require setting `TrustServerCertificate` to `true`.

https://github.com/npgsql/npgsql/blob/251fde9a55bdd78f378c0916d69f6944f68af995/src/Npgsql/Internal/NpgsqlConnector.cs#L547-L561

Reply:
The only reason we still have `TrustServerCertificate` is because we've changed the behavior of `SslMode.Require` (which doesn't check the certificate), so we throw an exception to make people aware of this and opt-int explicitly.

Reply:
Okay, so TrustServerCertificate serves a purpose of half measure arguments check. I would suggest to indicate that in the code then, because I've got missleaded.

Reply:
Kind of. Its original reason was to do exactly what is says: to trust the certificate in case of `SslMode.Require` (and maybe `SslMode.Prefer`, not sure about that). But with us adding `SslMode.VerifyCA` and `SslMode.VerifyFull`, it does become obsolete. So we just repurposed it, so as to
a) not break people using it with `SslMode.Require`
b) not add a new connection string parameter

We're going to remove `TrustServerCertificate` in Npgsql 7 or 8, but I guess it can't hurt adding a bit more info in the code...

Reply:
Requre SSL mode is currently also the only way to get custom certificate validation callback working.